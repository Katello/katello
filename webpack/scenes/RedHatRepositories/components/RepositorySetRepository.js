import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { ListView, Spinner, OverlayTrigger, Tooltip, Icon } from 'patternfly-react';
import { connect } from 'react-redux';

import { setRepositoryEnabled } from '../../../redux/actions/RedHatRepositories/repositorySetRepositories';
import '../index.scss';
import api from '../../../services/api';

class RepositorySetRepository extends Component {
  constructor(props) {
    super(props);

    this.state = { loading: false };

    this.setEnabled = (response) => {
      this.setState({ loading: false });

      const { data: { output: { repository: { id, name, content_type: type } } } } = response;

      const {
        productId, contentId, arch, releasever,
      } = this.props;

      const enabledRepo = {
        productId,
        contentId,
        id,
        name,
        type,
        arch,
        releasever,
      };

      this.props.setRepositoryEnabled(enabledRepo);
    };

    this.enableRepository = () => {
      this.setState({ loading: true });

      const {
        productId, contentId, arch, releasever,
      } = this.props;

      const url = `/products/${productId}/repository_sets/${contentId}/enable`;

      const data = {
        id: contentId,
        product_id: productId,
        basearch: arch,
        releasever: releasever || undefined,
      };

      api
        .put(url, data)
        .then(this.setEnabled)
        .catch(({ response: { data: error } }) => {
          this.setState({ loading: false, error });
        });
    };
  }

  render() {
    const { arch, releasever } = this.props;

    return (
      <ListView.Item
        heading={`${arch} ${releasever}`}
        className="list-item-with-divider"
        leftContent={
          this.state.error ? (
            <div className="list-error-danger">
              <Icon name="times-circle-o" />
            </div>
          ) : null
        }
        additionalInfo={
          this.state.error
            ? [
              <ListView.InfoItem key="error" stacked className="list-error-danger">
                {this.state.error.displayMessage}
              </ListView.InfoItem>,
              ]
            : null
        }
        actions={
          <Spinner loading={this.state.loading} inline>
            <OverlayTrigger
              overlay={<Tooltip id="enable">Enable</Tooltip>}
              placement="bottom"
              trigger={['hover', 'focus']}
              rootClose={false}
            >
              <button
                onClick={this.enableRepository}
                style={{
                  backgroundColor: 'initial',
                  border: 'none',
                  color: '#0388ce',
                }}
              >
                <i className={cx('fa-2x', 'fa fa-plus-circle')} />
              </button>
            </OverlayTrigger>
          </Spinner>
        }
        stacked
      />
    );
  }
}

RepositorySetRepository.propTypes = {
  contentId: PropTypes.number.isRequired,
  productId: PropTypes.number.isRequired,
  arch: PropTypes.string,
  releasever: PropTypes.string,
  setRepositoryEnabled: PropTypes.func.isRequired,
};

RepositorySetRepository.defaultProps = {
  releasever: '',
  arch: __('Unspecified'),
};

export default connect(null, { setRepositoryEnabled })(RepositorySetRepository);
