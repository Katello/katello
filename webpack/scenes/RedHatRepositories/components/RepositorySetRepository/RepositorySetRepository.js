import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { ListView, Spinner, OverlayTrigger, Tooltip, Icon, FieldLevelHelp } from 'patternfly-react';

import { sprintf, translate as __ } from 'foremanReact/common/I18n';
import { yStream } from '../RepositorySetRepositoriesHelpers';
import '../../index.scss';

class RepositorySetRepository extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  setEnabled = () => {
    this.props.setRepositoryEnabled(this.repoForAction());
  };

  repoForAction = () => {
    const {
      productId, contentId, arch, releasever, label,
    } = this.props;

    return {
      arch,
      productId,
      contentId,
      releasever,
      label,
    };
  };

  reloadEnabledRepos = () => (
    this.props.loadEnabledRepos({
      ...this.props.enabledPagination,
      search: this.props.enabledSearch,
    }, true)
  );

  notifyEnabled = (data) => {
    const repoName = data.output.repository.name;
    window.tfm.toastNotifications.notify({
      message: sprintf(__("Repository '%(repoName)s' has been enabled."), { repoName }),
      type: 'success',
    });
  };

  reloadAndNotify = async (result) => {
    if (result && result.success) {
      await this.reloadEnabledRepos();
      await this.setEnabled();
      await this.notifyEnabled(result.data);
    }
  };

  enableRepository = async () => {
    const result = await this.props.enableRepository(this.repoForAction());
    this.reloadAndNotify(result);
  };

  render() {
    const { displayArch, releasever, type } = this.props;

    const archLabel = displayArch || __('Unspecified');
    const releaseverLabel = releasever || '';

    const yStreamHelpText =
      sprintf(
        __('This repository is not suggested. Please see additional %(anchorBegin)sdocumentation%(anchorEnd)s prior to use.'),
        {
          anchorBegin: '<a href="https://access.redhat.com/articles/1586183">',
          anchorEnd: '</a>',
        },
      );
    // eslint-disable-next-line react/no-danger
    const yStreamHelp = <span dangerouslySetInnerHTML={{ __html: yStreamHelpText }} />;
    const shouldDeemphasize = () => type !== 'kickstart' && yStream(releaseverLabel);
    const repositoryHeading = () => (
      <span>
        {archLabel} {releaseverLabel}
        {shouldDeemphasize() ? (<FieldLevelHelp content={yStreamHelp} />) : null}
      </span>
    );

    return (
      <ListView.Item
        heading={repositoryHeading()}
        className={`list-item-with-divider ${shouldDeemphasize() ? 'deemphasize' : ''}`}
        leftContent={
          this.props.error ? (
            <div className="list-error-danger">
              <Icon name="times-circle-o" />
            </div>
          ) : null
        }
        additionalInfo={
          this.state.error
            ? [
              <ListView.InfoItem key="error" stacked className="list-error-danger">
                {this.state.error.displayMessage}
              </ListView.InfoItem>,
            ]
            : null
        }
        actions={
          <Spinner loading={this.props.loading} inline>
            <OverlayTrigger
              overlay={<Tooltip id="enable">Enable</Tooltip>}
              placement="bottom"
              trigger={['hover', 'focus']}
              rootClose={false}
            >
              <button
                onClick={this.enableRepository}
                style={{
                  backgroundColor: 'initial',
                  border: 'none',
                  color: '#0388ce',
                }}
              >
                <i className={cx('fa-2x', 'fa fa-plus-circle')} />
              </button>
            </OverlayTrigger>
          </Spinner>
        }
        stacked
      />
    );
  }
}

RepositorySetRepository.propTypes = {
  contentId: PropTypes.number.isRequired,
  productId: PropTypes.number.isRequired,
  displayArch: PropTypes.string,
  arch: PropTypes.string,
  releasever: PropTypes.string,
  type: PropTypes.string,
  label: PropTypes.string,
  enabledSearch: PropTypes.shape({
    query: PropTypes.string,
    searchList: PropTypes.string,
    filters: PropTypes.array,
  }),
  enabledPagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }).isRequired,
  loading: PropTypes.bool,
  error: PropTypes.bool,
  setRepositoryEnabled: PropTypes.func.isRequired,
  loadEnabledRepos: PropTypes.func.isRequired,
  enableRepository: PropTypes.func.isRequired,
};

RepositorySetRepository.defaultProps = {
  type: '',
  label: '',
  releasever: undefined,
  arch: undefined,
  displayArch: undefined,
  enabledSearch: {},
  loading: false,
  error: false,
};

export default RepositorySetRepository;
