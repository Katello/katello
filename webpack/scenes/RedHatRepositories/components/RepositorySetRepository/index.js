import { connect } from 'react-redux';
import { setRepositoryEnabled, enableRepository } from '../../../../redux/actions/RedHatRepositories/repositorySetRepositories';
import { loadEnabledRepos } from '../../../../redux/actions/RedHatRepositories/enabled';
import RepositorySetRepository from './RepositorySetRepository';

const mapStateToProps = (
  { katello: { redHatRepositories: { enabled } } },
  props,
) => ({
  ...props,
  enabledPagination: enabled.pagination,
  enabledSearch: enabled.search,
});

export default connect(
  mapStateToProps,
  { setRepositoryEnabled, loadEnabledRepos, enableRepository },
)(RepositorySetRepository);
