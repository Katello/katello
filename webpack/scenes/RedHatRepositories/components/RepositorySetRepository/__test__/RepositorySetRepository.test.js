import React from 'react';
import thunk from 'redux-thunk';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
import RepositorySetRepository from '../RepositorySetRepository';

jest.mock('../../../../../move_to_foreman/foreman_toast_notifications');

const mockStore = configureMockStore([thunk]);
const store = mockStore({});

describe('RepositorySetRepository Component', () => {
  let shallowWrapper;
  beforeEach(() => {
    shallowWrapper = shallow(<RepositorySetRepository
      store={store}
      contentId={1}
      productId={1}
      label="some label"
      displayArch="foo"
      releaseVer="1.1.1"
      type="foo"
      enabledPagination={{}}
      setRepositoryEnabled={() => {}}
      loadEnabledRepos={() => {}}
      enableRepository={() => {}}
    />);
  });

  afterEach(() => {
    store.clearActions();
  });

  it('should render', async () => {
    expect(toJson(shallowWrapper)).toMatchSnapshot();
  });
});
