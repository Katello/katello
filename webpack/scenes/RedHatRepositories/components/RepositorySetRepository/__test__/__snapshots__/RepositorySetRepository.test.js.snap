// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RepositorySetRepository Component should render 1`] = `
<ListViewItem
  actions={
    <Spinner
      className=""
      inline={true}
      inverse={false}
      loading={false}
      size="md"
    >
      <OverlayTrigger
        defaultOverlayShown={false}
        overlay={
          <Tooltip
            bsClass="tooltip"
            id="enable"
            placement="right"
          >
            Enable
          </Tooltip>
        }
        placement="bottom"
        rootClose={false}
        trigger={
          Array [
            "hover",
            "focus",
          ]
        }
      >
        <button
          onClick={[Function]}
          style={
            Object {
              "backgroundColor": "initial",
              "border": "none",
              "color": "#0388ce",
            }
          }
        >
          <i
            className="fa-2x fa fa-plus-circle"
          />
        </button>
      </OverlayTrigger>
    </Spinner>
  }
  additionalInfo={null}
  checkboxInput={null}
  className="list-item-with-divider "
  compoundExpand={false}
  compoundExpanded={false}
  description={null}
  heading={
    <span>
      foo
       
      
    </span>
  }
  hideCloseIcon={false}
  initExpanded={false}
  leftContent={null}
  onCloseCompoundExpand={[Function]}
  onExpand={[Function]}
  onExpandClose={[Function]}
  stacked={true}
/>
`;
